mockhub.go

```diff
@@ -0,0 +1,205 @@
+package frontman
+
+// a mock hub used in tests, similar to https://bitbucket.org/cloudradar/debug_hub/
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math/rand"
+	"net/http"
+	"strconv"
+)
+
+type MockHub struct {
+	address string
+}
+
+func NewMockHub(address string) *MockHub {
+	return &MockHub{
+		address: address,
+	}
+}
+
+func (hub *MockHub) URL() string {
+	return "http://" + hub.address
+}
+
+// returns some mocked checks
+func (hub *MockHub) indexHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		hub.getHandler(w, r)
+	case "POST":
+		hub.postHandler(w, r)
+	}
+}
+
+func (hub *MockHub) postHandler(w http.ResponseWriter, r *http.Request) {
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("MockHub: got post data", string(data))
+}
+
+func (hub *MockHub) getHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	query := r.URL.Query()
+	serviceChecksS := query.Get("serviceChecks")
+	webChecksS := query.Get("webChecks")
+
+	serviceChecks, err := strconv.ParseInt(serviceChecksS, 10, 64)
+	if err != nil {
+		serviceChecks = 5
+	}
+
+	webChecks, err := strconv.ParseInt(webChecksS, 10, 64)
+	if err != nil {
+		webChecks = 5
+	}
+
+	log.Println("responding with", serviceChecks, "serviceChecks and", webChecks, "webChecks")
+
+	checks := Input{
+		ServiceChecks: mockServiceChecks(int(serviceChecks)),
+		WebChecks:     mockWebChecks(int(webChecks)),
+	}
+
+	data, _ := json.Marshal(checks)
+	w.Write(data)
+}
+
+func (hub *MockHub) Serve() {
+
+	http.HandleFunc("/", hub.indexHandler)
+	log.Println("mock hub listening at", hub.URL())
+	http.ListenAndServe(hub.address, nil)
+}
+
+func mockServiceChecks(n int) []ServiceCheck {
+	res := []ServiceCheck{}
+	for i := 0; i < n; i++ {
+		res = append(res, randomServiceCheck())
+	}
+	return res
+}
+
+func mockWebChecks(n int) []WebCheck {
+	res := []WebCheck{}
+	for i := 0; i < n; i++ {
+		if i%5 == 0 {
+			// every fifth web check should be a lame web check
+			res = append(res, lameWebCheck())
+		} else {
+			res = append(res, randomWebCheck())
+		}
+	}
+	return res
+}
+
+func randomWebCheck() WebCheck {
+	methods := []string{"get", "post", "head"}
+	statuses := []int{200, 404}
+	patterns := []string{"running", "welcome", "yyy"}
+	return WebCheck{
+		UUID: randomUUID(),
+		Check: WebCheckData{
+			URL:                fmt.Sprintf("https://h%d.hostgum.eu/", rand.Intn(1000)),
+			Method:             methods[rand.Intn(len(methods))],
+			ExpectedHTTPStatus: statuses[rand.Intn(len(statuses))],
+			ExpectedPattern:    patterns[rand.Intn(len(patterns))],
+		},
+	}
+}
+
+func lameWebCheck() WebCheck {
+	return WebCheck{
+		UUID: randomUUID(),
+		Check: WebCheckData{
+			URL:                fmt.Sprintf("https://h1.hostgum.eu/sleep.php?t=%d", 5+rand.Intn(45)),
```

Comment:
host should be a configurable value

Reply:
No. This check is using a variable timeout on that backend and is a straight port of php script doing the same.

Reply:
sorry I mean host url

Reply:
Yea, the external script we rely on here lives on this URL. Try
```curl https://h1.hostgum.eu/sleep.php\?t\=5```

Reply:
We could at a later time port this to golang also so it's all self contained.